csrc/host_ir/jit.cpp

```diff
@@ -166,6 +178,401 @@ void printLlvmIr(llvm::Function* func, std::string_view msg) {
   llvm::outs() << "\n\n";
 }
 
+// Helper function to create value for binary operation
+llvm::Value* createValueForBinaryOp(BinaryOp* binary_op, std::unordered_map<Val*, llvm::Value*>& val_to_value, llvm::IRBuilder<>& builder) {
+  auto* lhs = binary_op->lhs()->as<Val>();
+  auto* rhs = binary_op->rhs()->as<Val>();
+  llvm::Value* lhs_value = getOrCreateValueForExtent(lhs, val_to_value, builder);
+  llvm::Value* rhs_value = getOrCreateValueForExtent(rhs, val_to_value, builder);
+  llvm::Value* out_value = nullptr;
+  if(binary_op->getBinaryOpType() == BinaryOpType::Add) {
+    out_value = builder.CreateAdd(lhs_value, rhs_value);
+  } else if(binary_op->getBinaryOpType() == BinaryOpType::Sub) {
+    out_value = builder.CreateSub(lhs_value, rhs_value);
+  } else if(binary_op->getBinaryOpType() == BinaryOpType::Mul) {
+    out_value = builder.CreateMul(lhs_value, rhs_value);
+  } else if(binary_op->getBinaryOpType() == BinaryOpType::CeilDiv) {
+    // Implement ceilDiv as (a + b - 1) / b
+    llvm::Value* numerator =
+    builder.CreateAdd(lhs_value, rhs_value);
+    llvm::Value* one = builder.getInt64(1);
+    numerator = builder.CreateSub(numerator, one);
+    out_value = builder.CreateUDiv(numerator, rhs_value);
+  } else {
+    NVF_THROW("LLVM Lowering Error: Unsupported binary operation type in extent calculation: ", binary_op->getBinaryOpType());
+  }
+  return out_value;
+}
+
+// Helper function to create value for unary operation
+llvm::Value* createValueForUnaryOp(UnaryOp* unary_op, std::unordered_map<Val*, llvm::Value*>& val_to_value, llvm::IRBuilder<>& builder) {
+  auto* in = unary_op->in()->as<Val>();
+  llvm::Value* in_value = getOrCreateValueForExtent(in, val_to_value, builder);
+  llvm::Value* out_value = nullptr;
+  if(unary_op->getUnaryOpType() == UnaryOpType::Cast) {
+    out_value = in_value;
+  } else if(unary_op->getUnaryOpType() == UnaryOpType::Abs) {
+    llvm::Value* is_negative = builder.CreateICmpSLT(in_value, builder.getInt64(0));
+    llvm::Value* negated = builder.CreateNeg(in_value);
+    out_value = builder.CreateSelect(is_negative, negated, in_value);
+  } else if(unary_op->getUnaryOpType() == UnaryOpType::Neg) {
+    out_value = builder.CreateNeg(in_value);
+  } else {
+    NVF_THROW("LLVM Lowering Error: Unsupported unary operation type in extent calculation: ", unary_op->getUnaryOpType());
+  }
+  return out_value;
+}
+// Helper function to create value for extent
+llvm::Value* createValueForExtent(
+    Val* val,
+    std::unordered_map<Val*, llvm::Value*>& val_to_value,
+    llvm::IRBuilder<>& builder) {
+  llvm::Value* out_value = nullptr;
+  if (val->isA<IterDomain>()) {
```

Comment:
Note for myself: it felt a bit strange that we kept an entry of IterDomain in this function. I think it's cleaner to have a evaluation of `Val*` directly in a function, rather than interleaving the handle of `IterDomain`, where we might want to handle extent differently.

Reply:
Agreed. I'd make `createValue`, `getOrCreateValue`, and `getOrCreateValueForExtent`. The first two take `Val*`, not `IterDomain*`. The third one takes `IterDomain*` and calls `getOrCreateValue` on the IterDomain's extent.